### DISCUSSION TITLE
(novice) Very frustrated trying to compile a "hello world" file.

### DISCUSSION THREAD

**Discussion by AidanofVT** [raysan5/raylib#2210] [Q&A]
My demo program: (I've tried others as well, including raylib's included demos) https://puu.sh/IwnpC/38766c0bb8.png

My file structure: https://puu.sh/IwniK/e98ca339bf.png

I followed the directions here: https://github.com/raysan5/raylib/wiki/Working-on-Windows
-- and in the video linked on that page, though that was intended for C. I confirmed that gcc is installed, and that raylib was "built" (I'm such a novice I don't really know what that means) in the raylib/src directory.
I've tried everything I can think of, but still get this output:

> PS C:\Users\Aidan\Desktop\C++_course\test> g++ source.cpp -o game.exe
C:\raylib\w64devkit\bin/ld.exe: C:\Users\Aidan\AppData\Local\Temp\ccK4IEjg.o:source.cpp:(.text+0x17): undefined reference to `SetTargetFPS'
C:\raylib\w64devkit\bin/ld.exe: C:\Users\Aidan\AppData\Local\Temp\ccK4IEjg.o:source.cpp:(.text+0x33): undefined reference to `InitWindow'
C:\raylib\w64devkit\bin/ld.exe: C:\Users\Aidan\AppData\Local\Temp\ccK4IEjg.o:source.cpp:(.text+0x3a): undefined reference to `BeginDrawing'
C:\raylib\w64devkit\bin/ld.exe: C:\Users\Aidan\AppData\Local\Temp\ccK4IEjg.o:source.cpp:(.text+0x62): undefined reference to `ClearBackground'
C:\raylib\w64devkit\bin/ld.exe: C:\Users\Aidan\AppData\Local\Temp\ccK4IEjg.o:source.cpp:(.text+0xa5): undefined reference to `DrawText'
C:\raylib\w64devkit\bin/ld.exe: C:\Users\Aidan\AppData\Local\Temp\ccK4IEjg.o:source.cpp:(.text+0xaa): undefined reference to `EndDrawing'
C:\raylib\w64devkit\bin/ld.exe: C:\Users\Aidan\AppData\Local\Temp\ccK4IEjg.o:source.cpp:(.text+0xaf): undefined reference to `WindowShouldClose'
>collect2.exe: error: ld returned 1 exit status
 
Clearly there's some sort of issue with the compiler finding its way to the raylib libraries, but I'm out of ideas as to how to resolve it. Please help. I'm trying to learn C++ more than raylib specifically, so it's important to me that I really understand what's going wrong here.

**Comment by raysan5** [comment]
@AidanofVT You need to link with the required libraries:
```
g++ source.cpp -o game.exe -O2 -std=c99 -Wall -I$(RAYLIB_PATH)/src/external -I$(RAYLIB_PATH)/src -I$(RAYLIB_PATH)/src/extras -DPLATFORM_DESKTOP -lraylib -lopengl32 -lgdi32 -lwinmm
```
I recommend you investigate and learn how `g++` compiler works and the required parameters to compile an executable from your code file.

I provide a preconfigured raylib installer with Notepad++ preconfigured for beginners to start coding with raylib. For other IDEs and configurations, it requires understanding how a build system and a compiler works.

**Reply by AidanofVT** [reply]
Thank you, that worked for me. I've researched the compiler's usage, and I have a couple questions:

1. When I include "-std=c99", I get the warning: "command-line option '-std=c99' is valid for C/ObjC but not for C++." Why include it?
2. src/external and src/extras are subdirectories of src. It seems like the "-I" parameter searches both the given directory and its subdirectories, so why not just keep it to "-I$(RAYLIB_PATH)/src"?

**Reply by raysan5** [reply]
1. That's right, I just copied the code from `gcc`, it's not required with C++.
2. Because subdirectories are not included afaik, you need to specify all the required paths.

**Reply by orcmid** [reply]
2. I know some systems accept `directory/**` to signify the directory and all of its subdirectories.  Might be rather platform and compiler/tool specific.  Specifying the subdirectors for includes is safe and rarely harmful.
 1. The specification of c99 (or c11 or c17) is meaningful for some C++ compilers and valuable for the *.h and *.c files.  Since @AidanofVT is compiling source.cpp, I believe that is the conflict, not that gcc and g++ won't handle the Standard C cases when given a *.c file to compile (or an option to treat all files as c.  

@AidanofVT, as a novice, try using source.c instead.  I presume that you are *not* attempting to use C++ features that are not in common with clean C language.  For starting out, use C language for all of your raylib exploration.  It will provide solid grounding for more-elaborate project, especially understanding how the compilation process works and how to deal with error reports and making it easy to find your own mistakes beyond what the compilers and linkers check. Once you've mastered that level, you can begin to fly from the nest.

**Reply by AidanofVT** [reply]
Wow, @raysan5, you sure are an attentive respondent! You got back to me like ninety seconds after I posted those questions. Now here's one more: since I'm pointing the compiler to the raylib header file in raylib/src, why do I still need to have the _#include "raylib.h"_ pre-processing directive in my source file? What's the difference between _#include_ given to the compiler at the top of the source file and _-I_ given to the compiler (or would it be the linker) in the terminal command?

**Reply by orcmid** [reply]
> What's the difference between _#include_ given to the compiler at the top of the source file and _-I_ given to the compiler (or would it be the linker) in the terminal command?

@AidanofVT  `#include "raylib.h"` says what (header) file you want, it doesn't tell the compiler where to find it if it is not in the same directory as the source-code that has the '#include'.  The `-I` options and their counterparts (including environment variables) specify other places worth looking into for resolving `#include` references.  (The form `include <raylib.h>` signifies that only other places should be searched.)

Something similar is done in finding a code library that might resolve external references from the code.  Don't think about that too much at this point.   It is different than #include in C/C++ and involves a linker for combining previously-compiled code with your freshly-compiled code to make the executable.

**Reply by AidanofVT** [reply]
Got it, thank you.

**Reply by caiubyfreitas** [reply]
Have you tried this:

cc main.c -lraylib -lGL -lm -lpthread -ldl -lrt -lX11
